### ISSUE TITLE
cmd/compile: misleading error when supplying a statement as a function parameter

### ISSUE THREAD

**Issue by 0xmohit** [golang/go#15307]
```
$  go version
go version go1.6 linux/amd64
```

```
$  go env
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH=""
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GO15VENDOREXPERIMENT="1"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0"
CXX="g++"
CGO_ENABLED="1"
```

Tried to compile the following:

``` go
package main

import "fmt"

func main() {
        i := 42
        fmt.Println((i++))
}
```

```
t.go:7: syntax error: unexpected ++, expecting )
t.go:8: syntax error: non-declaration statement outside function body
```

The second error should be suppressed.  (It's not observed using 1.5.3 and earlier.)

**Comment by ALTree** [comment]
Not limited to function parameters:

```
package main

func main() {
	i := 42
	(i++)
}
```

```
$ go run prova.go
# command-line-arguments
./prova.go:5: syntax error: unexpected ++, expecting )
./prova.go:6: syntax error: non-declaration statement outside function body
```

Also

```
package main

func main() {
	i := 42
	(var int)
}
```

gives the same double error. Seems like the parser is confused by certain syntax issues when inside `_{L,R}paren`.

**Comment by griesemer** [comment]
Except for grouping of declarations, inside ()'s we can only have expressions. Possibly error recovery could be a bit stronger in that case.

**Comment by griesemer** [comment]
It's tricky to come up with a good parser recovery strategy in cases like these. The culprit is the `syntax.parser.want` method:
```
 func (p *parser) want(tok token) {
        if !p.got(tok) {
                p.syntax_error("expecting " + tok.String())
                p.advance()
        }
 }
```
If a `++` (or any other token) is found instead of the closing token with `tok` == `)`, an error is reported and the offending token is consumed. This assumes that the expected token is simply missing. In these cases, the expected `)` follows. So a simple way to work around this here is to change the `want` method as follows:
```
@@ -90,7 +90,8 @@ func (p *parser) got(tok token) bool {
 func (p *parser) want(tok token) {
        if !p.got(tok) {
                p.syntax_error("expecting " + tok.String())
-               p.advance()
+               p.advance(tok)
+               p.got(tok) // try once more
        }
 }
```
This will fix these issues, but cause others (run the tests).

Leaving open for further exploration. Not urgent.

**Comment by 0xmohit** [comment]
This appears to be fixed in c37090f00f57b38b4c8b93b1d5e97f5a10cb1765.

**Comment by griesemer** [comment]
Confirming. Thanks for noticing this, @0xmohit .
